### PR TITLE
feat: add simple login

### PR DESCRIPTION
**Pull request by haveachin**

### CODE REVIEW

File: auth/src/lib.rs
Comment (nitnelave) [inline auth/src/lib.rs:54]: Server reset? What's that for? Maybe we need a better name.

File: auth/src/lib.rs
Comment (nitnelave) [inline auth/src/lib.rs:92]: I'd like to split this PR into login (that I would more easily approve) and registration/reset, that I would like to keep OPAQUE only if possible. For me, the easy login method is not meant to be opened up to the world, but more like running scripts on the same host more conveniently.

File: server/src/infra/auth_service.rs
Comment (nitnelave) [inline server/src/infra/auth_service.rs:67]: Instead of a nested match none, you can use a method on option to combine the 2. I'm not sure exactly which method, but something like `or_else` would probably help.

Of course, you'd need to map them to the same type, some kind of string.
If it gets too complicated, it's fine like this.

File: auth/src/lib.rs
Comment (nitnelave) [inline auth/src/lib.rs:50]: Maybe `Option<String>`?

File: auth/src/lib.rs
Comment (nitnelave) [inline auth/src/lib.rs:40]: Which convention? :)

Maybe better to rename to simple http login

File: server/src/infra/auth_service.rs
Comment (nitnelave) [inline server/src/infra/auth_service.rs:340]: I don't think you need this clone. You can write `let password = & request.password;`

File: server/src/infra/auth_service.rs
Comment (nitnelave) [inline server/src/infra/auth_service.rs:333]: Are there no refactoring opportunities for the 2 places that do the entire protocol on the server side?
Maybe returning a result with a custom enum as error for the various cases, cleaning up the code and separating the logic and the error handling.

File: server/src/infra/auth_service.rs
Comment (nitnelave) [inline server/src/infra/auth_service.rs:458]: As mentioned above, let's split that into another PR

File: server/src/infra/auth_service.rs
Comment (nitnelave) [inline server/src/infra/auth_service.rs:662]: Maybe `/login/simple`?

**Review by nitnelave — CHANGES_REQUESTED** [review]
Thanks for the PR!
Overall, this looks good, it has all the required bits and bobs and probably works.
It also hopelessly breaks the client since it cannot parse the login response, so I can't merge it until the client changes are there too :)

File: auth/src/lib.rs
Comment (haveachin) [reply auth/src/lib.rs:39]: Not sure what you mean by that. I don't see any message structs with masking secrets. Do you mean the triple-slash comments?

Comment (nitnelave) [reply auth/src/lib.rs:39]: Ah, sorry, it's only in the dependencies. See https://github.com/kanidm/ldap3_proto/pull/6/files for example.

File: auth/src/lib.rs
Comment (haveachin) [reply auth/src/lib.rs:92]: So you would basically require any service that is automatically registering users to use OPAQUE?

File: server/src/infra/auth_service.rs
Comment (haveachin) [reply server/src/infra/auth_service.rs:67]: Sorry, I tried it for like 2 hours and I could figure it out. I'm sure that there is a way, but I'm most certain that it will be more complex that what I have now.

File: server/src/infra/auth_service.rs
Comment (haveachin) [reply server/src/infra/auth_service.rs:333]: I got no clue how Rust works, but I remember something about error handling in one Rust talk I watched. I will try something, but no guarantee that is any good.

File: server/src/infra/auth_service.rs
Comment (haveachin) [reply server/src/infra/auth_service.rs]: I changed the routes to `/simple/login` and `/simple/register`

File: server/src/infra/auth_service.rs
Comment (haveachin) [reply server/src/infra/auth_service.rs]: fixed

File: auth/src/lib.rs
Comment (haveachin) [reply auth/src/lib.rs]: Renamed it to password_reset::ServerPasswordResetResponse

File: auth/src/lib.rs
Comment (haveachin) [reply auth/src/lib.rs]: Changed it

File: auth/src/lib.rs
Comment (haveachin) [reply auth/src/lib.rs]: Changed it to ClientSimpleLoginRequest

File: server/src/infra/auth_service.rs
Comment (haveachin) [reply server/src/infra/auth_service.rs]: That really sucks, because I actually need to register users via API without the opaque protocol

File: server/src/infra/auth_service.rs
Comment (haveachin) [reply server/src/infra/auth_service.rs]: I got something working, but I'm not sure if it is better than what I had before

Comment (haveachin) [reply server/src/infra/auth_service.rs:71]: nvm, the str pointer were already empty when I was trying to read them

File: app/src/components/login.rs
Comment (nitnelave) [inline app/src/components/login.rs:90]: I think you don't need the lambda here, you can leave it as it was (a reference to the enum constructor)

File: app/src/components/login.rs
Comment (nitnelave) [inline app/src/components/login.rs:95]: Can you use longer, more explicit names than r, b? I'm reviewing the PR and have already lost context on what they represent

File: app/src/infra/api.rs
Comment (nitnelave) [inline app/src/infra/api.rs:195]: Hmm, sorry I'm reviewing on my phone so it's a bit hard for me to tell exactly what's going on here, but:
- you get a Result
- you add more info to the error with context()
- if it was no error, you set the cookies
- and then you *throw away the error* and just return whether it was successful (is_ok)

I'll have a closer look later, why this function returns a bool, but I'd like to avoid throwing away errors.

File: server/src/infra/auth_service.rs
Comment (nitnelave) [inline server/src/infra/auth_service.rs:1]: Is that cargo fmt reordering the imports? Maybe we should fuse the various std imports. Small nit, feel free to ignore.

File: server/src/infra/auth_service.rs
Comment (nitnelave) [inline server/src/infra/auth_service.rs:348]: Empty error string?

File: auth/src/lib.rs
Comment (nitnelave) [reply auth/src/lib.rs]: Let's move this conversation to the next PR, since I'd like to anyway split it into one for login and one for registration.

File: server/src/infra/auth_service.rs
Comment (nitnelave) [reply server/src/infra/auth_service.rs]: I'm not saying no, I'm saying that we'll talk about it in the next PR :) I can probably easily be convinced.

**Review by nitnelave — CHANGES_REQUESTED** [review]
Alright, that looks much better! Still a couple of nits, and I'll have a last check once I'm on my computer, but it looks good now.

File: server/src/infra/auth_service.rs
Comment (haveachin) [reply server/src/infra/auth_service.rs:1]: Not sure, I just formatted the file using rustfmt

File: app/src/components/login.rs
Comment (haveachin) [reply app/src/components/login.rs]: I refactored it to a lambda

File: app/src/components/login.rs
Comment (haveachin) [reply app/src/components/login.rs]: They should be something like response and success (?), but i'm not 100% sure what the bool is

File: app/src/infra/api.rs
Comment (haveachin) [reply app/src/infra/api.rs:195]: Yeah, the problem is that it expects a result with a Result<(login::ServerLoginResponse, bool) and I'm not sure how I should change. I'm actually happy that I got it working. I find it really hard to read Rust code in general, sorry :(

File: server/src/infra/auth_service.rs
Comment (haveachin) [reply server/src/infra/auth_service.rs]: It returns now "Internal Server Error: " appended with the error

File: server/src/infra/auth_service.rs
Comment (haveachin) [reply server/src/infra/auth_service.rs]: OK, I moved it to a separate branch

File: app/src/infra/api.rs
Comment (nitnelave) [reply app/src/infra/api.rs:195]: Okay, I checked out your branch to have a better look, and I see the issue. The String returned here is the username, not the token. You were right in parsing the server login response and _then_ calling `get_claims_from_jwt`, but you can just stop after `and_then(set_cookies)` like the original code. The callback still takes a String

Comment (haveachin) [reply app/src/infra/api.rs]: Oof, I overcomplicated it. Should be fixed now

File: server/src/infra/auth_service.rs
Comment (nitnelave) [reply server/src/infra/auth_service.rs:71]: How about:
```rust
    match (
      request.cookie("refresh_token"),
      request.headers().get("refresh-token")
    ) {
      (Some(c), _) => parse_refresh_token(c.value()),
      (_, Some(t)) => parse_refresh_token(t.to_str().unwrap()),
      (None, None) => Err(HttpResponse::Unauthorized().body("Missing refresh token")),
```

I think it captures well the logic.

**Review by nitnelave — CHANGES_REQUESTED** [review]

**Comment by nitnelave** [comment]
Thanks for your efforts, we're almost there!

File: server/src/infra/auth_service.rs
Comment (nitnelave) [inline server/src/infra/auth_service.rs:346]: `password`, not `&password`

File: auth/src/lib.rs
Comment (nitnelave) [reply auth/src/lib.rs]: I didn't mean to _derive_ Debug, but to implement it by hiding the password. Nevermind, I'll do it myself after this PR.

**Review by nitnelave — APPROVED** [review]

**Comment by nitnelave** [comment]
One last thing before I merge: can you squash your commits? Or even better organize them into logical independent units that each pass the tests, but I think for this PR it's enough to squash.

File: auth/src/lib.rs
Comment (nitnelave) [inline auth/src/lib.rs:49]: You're just missing `.field("password", "***********")` and you're golden ;)

**Review by nitnelave — CHANGES_REQUESTED** [review]

File: auth/src/lib.rs
Comment (nitnelave) [reply auth/src/lib.rs:49]: (and re-squash the commits)

Comment (haveachin) [reply auth/src/lib.rs:49]: done

**Review by nitnelave — APPROVED** [review]

**Comment by nitnelave** [comment]
Thanks a lot for the PR! And for being patient with the nits :)

**Pull request merged**
